src/postprocess/dos.jl

```diff
@@ -67,3 +67,120 @@ end
+            for n_spin = 1 : basis.model.n_spin_components
+                pdos[n_spin][i,j] = pdos_ij[n_spin]
+            end
+        end
+    end
+
+   pdos
+end
+
+function compute_pdos(ε, i::Integer, l::Integer,
+    iatom::Integer, basis, eigenvalues, ψ; kwargs...)
+    compute_pdos(ε, basis, eigenvalues, ψ, i, l, basis.model.atoms[iatom].psp, basis.model.positions[iatom]; kwargs...)
+end
+
+function compute_pdos(i::Integer, l::Integer, iatom::Integer, scfres::NamedTuple;
+    ε=scfres.εF, kwargs...)
+    scfres = unfold_bz(scfres)
+    compute_pdos(ε, i, l, iatom, scfres.basis, scfres.eigenvalues, scfres.ψ; kwargs...)
+end
+
+
+function compute_pdos_projs(basis, eigenvalues, ψ, i::Integer, l::Integer, psp, position)
+
+    projs = Vector{Matrix}(undef, length(eigenvalues))
+    position = vector_red_to_cart(basis.model, position)
+
+    G_plus_k_all = [to_cpu(Gplusk_vectors_cart(basis, basis.kpoints[ik])) for ik = 1:length(basis.kpoints)]
+    fourier_form = atom_fourier_form(i,l,psp,G_plus_k_all)
+
+    for (ik, ψk) in enumerate(ψ)
+        fourier_form_k = fourier_form[ik]
+        atom_shift = [dot(position, Gi) for Gi in G_plus_k_all[ik]]
+        atom_fourier_k = exp.(-im * atom_shift) .* fourier_form_k ./ sqrt(basis.model.unit_cell_volume)
+        ak_normalize = similar(atom_fourier_k)
+        for m = 1:2l+1
+            ak_normalize[:, m] = atom_fourier_k[:, m] ./ norm(atom_fourier_k[:, m])
+        end
+        projs[ik] = abs2.(ψk' * ak_normalize)
+    end
+
+    projs
+end
+
+
+"""
+Build form factors (Fourier transforms of pseudo-atomic wavefunctions) for an atom centered at 0.
+"""
+function atom_fourier_form(i::Integer, l::Integer, psp, G_plus_k_all::Vector{Vector{Vec3{TT}}}) where {TT}
+    T = real(TT)
+    @assert psp isa PspUpf
+    # Pre-compute the radial parts of the  pseudo-atomic wavefunctions at unique |p| to speed up
+    # the form factor calculation (by a lot). Using a hash map gives O(1) lookup.
+    
+    radials = IdDict{T,T}()  # IdDict for Dual compatibility
+    for (ik, G_plus_k) in enumerate(G_plus_k_all)
+        for p in G_plus_k
+            p_norm = norm(p)
+            if !haskey(radials, p_norm)
+                radials[p_norm] = eval_psp_pswfc_fourier(psp, i, l, p_norm)
+            end
+        end
+    end
+
+    form_factors = Vector{Matrix{Complex{T}}}(undef, length(G_plus_k_all))
+    for (ik, G_plus_k) in enumerate(G_plus_k_all)
+        form_factors_ik = Matrix{Complex{T}}(undef, length(G_plus_k), 2l + 1)
+        for (ip, p) in enumerate(G_plus_k)
+            radials_p = radials[norm(p)]
+            count = 1
+            for m = -l:l
+                # see "Fourier transforms of centered functions" in the docs for the formula
+                angular = (-im)^l * ylm_real(l, m, p)
+                form_factors_ik[ip, m+l+1] = radials_p * angular
+            end
+        end
+        form_factors[ik] = form_factors_ik
+    end
+    form_factors
+end
```

Comment:
that's a lot of copy and paste with the non-local projector code of the pseudos. Should we try to unify the code pieces. In any case you should make a comment to indicate future possibilities for refactoring here and in the non-local projector code of the pseudos.